packages/aws-cdk/lib/cxapp/cloud-executable.ts

```diff
@@ -82,51 +83,64 @@ export class CloudExecutable implements ICloudAssemblySource {
+              + `Missing context keys: '${Array.from(missingKeys).join(', ')}'`);
+          }
 
-      if (assembly.manifest.missing && assembly.manifest.missing.length > 0) {
-        const missingKeys = missingContextKeys(assembly.manifest.missing);
+          let tryLookup = true;
+          if (previouslyMissingKeys && setsEqual(missingKeys, previouslyMissingKeys)) {
+            await this.props.ioHelper.defaults.debug('Not making progress trying to resolve environmental context. Giving up.');
+            tryLookup = false;
+          }
 
-        if (!this.canLookup) {
-          throw new ToolkitError(
-            'Context lookups have been disabled. '
-            + 'Make sure all necessary context is already in \'cdk.context.json\' by running \'cdk synth\' on a machine with sufficient AWS credentials and committing the result. '
-            + `Missing context keys: '${Array.from(missingKeys).join(', ')}'`);
-        }
+          previouslyMissingKeys = missingKeys;
 
-        let tryLookup = true;
-        if (previouslyMissingKeys && setsEqual(missingKeys, previouslyMissingKeys)) {
-          await this.props.ioHelper.defaults.debug('Not making progress trying to resolve environmental context. Giving up.');
-          tryLookup = false;
-        }
+          if (tryLookup) {
+            await this.props.ioHelper.defaults.debug('Some context information is missing. Fetching...');
 
-        previouslyMissingKeys = missingKeys;
+            const updates = await contextproviders.provideContextValues(
+              assembly.manifest.missing,
+              this.props.sdkProvider,
+              GLOBAL_PLUGIN_HOST,
+              this.props.ioHelper,
+            );
 
-        if (tryLookup) {
-          await this.props.ioHelper.defaults.debug('Some context information is missing. Fetching...');
+            for (const [key, value] of Object.entries(updates)) {
+              this.props.configuration.context.set(key, value);
+            }
 
-          const updates = await contextproviders.provideContextValues(
-            assembly.manifest.missing,
-            this.props.sdkProvider,
-            GLOBAL_PLUGIN_HOST,
-            this.props.ioHelper,
-          );
+            // Cache the new context to disk
+            await this.props.configuration.saveContext();
 
-          for (const [key, value] of Object.entries(updates)) {
-            this.props.configuration.context.set(key, value);
+            // Execute again
+            continue;
           }
-
-          // Cache the new context to disk
-          await this.props.configuration.saveContext();
-
-          // Execute again
-          continue;
         }
+        return new CloudAssembly(assembly, this.props.ioHelper);
       }
+    } catch (e: any) {
+      error = e;
+    } finally {
+      await synthSpan.end({ error });
 
-      return new CloudAssembly(assembly, this.props.ioHelper);
+      if (error) {
+        throw error;
+      }
     }
```

Comment:
quibble: this doesn't ready nice at all

if you want to different kind of endings, why not call `synthSpan.end()` twice: Once in the try block and once in the catch?